source/matrix_free/vector_data_exchange.cc

```diff
@@ -316,6 +318,965 @@ namespace internal
+          for (unsigned int i = 0; i < recv_sm_ranks.size(); i++)
+            MPI_Irecv(recv_sm_indices.data() + recv_sm_ptr[i],
+                      recv_sm_ptr[i + 1] - recv_sm_ptr[i],
+                      MPI_UNSIGNED,
+                      recv_sm_ranks[i],
+                      2,
+                      comm_sm,
+                      recv_sm_req.data() + i);
+
+          MPI_Waitall(recv_sm_req.size(),
+                      recv_sm_req.data(),
+                      MPI_STATUSES_IGNORE);
+          MPI_Waitall(send_sm_req.size(),
+                      send_sm_req.data(),
+                      MPI_STATUSES_IGNORE);
+        }
+
+        {
+          send_sm_offset.resize(send_sm_ranks.size());
+
+          for (unsigned int i = 0; i < send_sm_ranks.size(); i++)
+            MPI_Irecv(send_sm_offset.data() + i,
+                      1,
+                      MPI_UNSIGNED,
+                      send_sm_ranks[i],
+                      3,
+                      comm_sm,
+                      send_sm_req.data() + i);
+
+          for (unsigned int i = 0; i < recv_sm_ranks.size(); i++)
+            MPI_Isend(recv_sm_offset.data() + i,
+                      1,
+                      MPI_UNSIGNED,
+                      recv_sm_ranks[i],
+                      3,
+                      comm_sm,
+                      recv_sm_req.data() + i);
+
+          MPI_Waitall(recv_sm_req.size(),
+                      recv_sm_req.data(),
+                      MPI_STATUSES_IGNORE);
+          MPI_Waitall(send_sm_req.size(),
+                      send_sm_req.data(),
+                      MPI_STATUSES_IGNORE);
+        }
+
+        internal::compress(recv_sm_ptr, recv_sm_indices, recv_sm_len);
+
+        internal::compress(send_remote_ptr,
+                           send_remote_indices,
+                           send_remote_len);
+        send_remote_offset.clear();
+        send_remote_offset.push_back(0);
+
+        for (unsigned int r = 0, c = 0; r < send_remote_ranks.size(); r++)
+          {
+            for (unsigned int i = send_remote_ptr[r];
+                 i < send_remote_ptr[r + 1];
+                 i++)
+              c += send_remote_len[i];
+            send_remote_offset.push_back(c);
+          }
+
+        internal::compress(send_sm_ptr, send_sm_indices, send_sm_len);
+        internal::compress(recv_sm_ptr_, recv_sm_indices_, recv_sm_len_);
+        internal::compress(send_sm_ptr_, send_sm_indices_, send_sm_len_);
+
+#endif
+      }
+
+
+
+      void
+      Full::export_to_ghosted_array_start(
+        const unsigned int                          communication_channel,
+        const ArrayView<const double> &             locally_owned_array,
+        const std::vector<ArrayView<const double>> &shared_arrays,
+        const ArrayView<double> &                   ghost_array,
+        const ArrayView<double> &                   temporary_storage,
+        std::vector<MPI_Request> &                  requests) const
+      {
```

Comment:
I must say that I do not particularly like to have hardcoded functions for both `double` and `float` because it would make extensions to say `std::complex<double>` (if that ever makes sense) cumbersome and double the number of functions. I realize it is the "best" alternative behind a `virtual` function interface for a class that we do not want to be templated by `Number`, a requirement to be able to run `MF::loop` on vectors of different type than the `MatrixFree` type on its own. If in the end we could move all code through a single realization of `VectorDataExchange` then of course we could skip the virtual functions and go to the templated, aka `_impl` path, at the expense of moving everything into the header file. Anyway, if anyone has other suggestions it would be good to hear them.

Reply:
> I must say that I do not particularly like to have hardcoded functions for both double and float

But I have no other solution...